creusot-contracts/src/ghost/ptr_own.rs

```diff
@@ -234,17 +229,194 @@ impl<T: ?Sized> PtrOwn<T> {
+    #[ensures(result.len() == self.len())]
+    #[ensures(forall<i> 0 <= i && i < self.len()
+        ==> result[i].ptr() == self.ptr().thin().offset_logic(i)
+        && *result[i].val() == self.val()@[i])]
+    #[ensures((^self).ptr() == self.ptr())]
+    // NOTE: If the caller swaps `result[i]` with another token, there are no guarantees about the final value in `self`.
+    #[ensures(forall<i> 0 <= i && i < self.len()
+        && result[i].ptr() == (^result[i]).ptr()
+        ==> *(^result[i]).val() == (^self).val()@[i])]
+    pub fn elements_mut(&mut self) -> Seq<&mut PtrOwn<T>> {
+        panic!("called ghost function in normal code")
+    }
+
+    /// Convert a `&PtrOwn<[T]>` for a non-empty slice into a `&PtrOwn<T>` for the element at the given index.
+    #[check(ghost)]
+    #[requires(0 <= index && index < self.len())]
+    #[ensures(result.ptr() == self.ptr().thin().offset_logic(index))]
+    #[ensures(*result.val() == self.val()@[index])]
+    pub fn index_ptr_own_ref_ghost(&self, index: Int) -> &PtrOwn<T> {
+        let mut r = self.elements();
+        r.split_off_ghost(index).pop_front_ghost().unwrap()
+    }
+
+    /// Convert a `&mut PtrOwn<[T]>` for a non-empty slice into a `&mut PtrOwn<T>` for the element at the given index.
+    #[check(ghost)]
+    #[requires(0 <= index && index < self.len())]
+    #[ensures(result.ptr() == self.ptr().thin().offset_logic(index))]
+    #[ensures(*result.val() == self.val()@[index])]
+    #[ensures((^self).ptr() == self.ptr())]
+    // NOTE: If the caller swaps `result` with another token, there are no guarantees about the final value in `self`.
+    #[ensures(result.ptr() == (^result).ptr() ==> *(^result).val() == (^self).val()@[index])]
+    #[ensures(forall<k: Int> k != index ==> (^self).val()@.get(k) == self.val()@.get(k))]
+    pub fn index_ptr_own_mut_ghost(&mut self, index: Int) -> &mut PtrOwn<T> {
+        let mut r = self.elements_mut();
+        r.split_off_ghost(index).pop_front_ghost().unwrap()
+    }
+
+    /// Convert `&PtrOwn<[T]>` to `&PtrLive<T>`
+    #[trusted]
+    #[check(ghost)]
+    #[ensures(result.ptr() == self.ptr().thin())]
+    #[ensures(result.len()@ == self.len())]
+    pub fn live(&self) -> &PtrLive<'_, T> {
+        panic!("called ghost function in normal code")
+    }
+
+    /// Split a `&PtrLive<T>` token out of `&mut PtrOwn<[T]>`.
+    ///
+    /// The original `&mut PtrOwn<T>` is returned so that
+    /// it can be used even during the lifetime of the `&PtrLive`.
+    #[trusted]
+    #[check(ghost)]
+    #[ensures(*result.0 == *self)]
+    #[ensures(result.1.ptr() == self.ptr().thin())]
+    #[ensures(result.1.len()@ == self.len())]
+    #[ensures((^self).ptr() == self.ptr())]
+    // NOTE: If the caller swaps `result.0` with another token, there are no guarantees about the final value.
+    #[ensures((^result.0).ptr() == result.0.ptr() ==> (^result.0).val() == (^self).val())]
+    pub fn live_mut(&mut self) -> (&mut Self, &PtrLive<'_, T>) {
```

Comment:
```suggestion
    pub fn live_mut(&mut self) -> (&mut Self, PtrLive<T>) {
```

No need to use borrows of `PtrLive`, and (I think that) `'_, ` is implicit.